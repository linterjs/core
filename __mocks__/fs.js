const realFs = jest.requireActual("fs");
const { vol } = jest.requireActual("memfs");
const { ufs } = jest.requireActual("unionfs");
ufs.use(realFs).use(vol);
module.exports = ufs;
