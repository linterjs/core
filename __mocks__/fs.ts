const fs = require.requireActual("fs");

let files = Object.create(null);

function __clearFiles() {
  files = Object.create(null);
}

function __setFiles(newFiles: object) {
  Object.entries(newFiles).forEach(([path, content]) => {
    files[path] = content;
  });
}

const mockFs = {
  ...fs,
  __clearFiles,
  __setFiles,
  readFileSync: jest.fn((path, options) => {
    if (path in files) {
      return files[path];
    }

    fs.readFileSync(path, options);
  }),
  writeFileSync: jest.fn((path, data, options) => {
    if (path in files) {
      files[path] = data;
    } else {
      fs.writeFileSync(path, data, options);
    }
  })
};

module.exports = mockFs;
