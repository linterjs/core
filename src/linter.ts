import { createFormat, FormatFunction } from "./format";
import { createLint, LintFunction } from "./lint";
import { LinterAdapter, loadedLinterAdapterPromises } from "./linter-adapter";
import { loadLinterProvidersFromFile } from "./linter-providers";

// XXX: Create a new instance of Linter to load/reload registered
// linter providers from file
export default class Linter {
  format: FormatFunction = createFormat(loadedLinterAdapterPromises.get(this));
  lint: LintFunction = createLint(loadedLinterAdapterPromises.get(this));

  // TODO: Support config/options
  constructor() {
    const linterProviders = loadLinterProvidersFromFile();
    const linterAdapters = new Set<Promise<LinterAdapter>>();
    for (const { linterFactory } of linterProviders.values()) {
      linterAdapters.add(Promise.resolve(linterFactory()));
    }
    loadedLinterAdapterPromises.set(this, linterAdapters);
  }
}
