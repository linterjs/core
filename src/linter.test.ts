import { __clearFiles, __setFiles } from "fs";
import { Linter } from "./linter";
import {
  registerLinterProvider,
  registeredLinterProvidersFilePath
} from "./linter-providers";

jest.mock("@linter/eslint");
jest.mock("fs");

beforeAll(() => {
  __setFiles({
    [registeredLinterProvidersFilePath]: ""
  });
});

describe("Create Linter", () => {
  test("without registered linter providers", () => {
    expect(() => {
      const linter = new Linter();
    }).toThrowErrorMatchingSnapshot();
  });

  test("with @linter/eslint registered", () => {
    registerLinterProvider("@linter/eslint");
    const linter = new Linter();
    expect(linter).toBeInstanceOf(Linter);
  });
});
