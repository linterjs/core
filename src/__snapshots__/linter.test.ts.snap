// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Linter without registered linter providers 1`] = `"No linters registered. Please install at least one linter provider package."`;
